gitpod-io/gitpod
# [ws-proxy] Decouple ws-proxy from ws-manager

## Description

This refactoring removes the use of services to export workspace ports, using the pod directly instead.
Changing the default behavior, we do not rely on DNS to route traffic to the workspace (ws-<workspaceID>.<ns>.cluster.local:<port>) anymore, and instead, we use the pod IP address. This removes issues related to DNS queries, timeous, or UDP connection tracking (iptables).
A side effect of this change is kube-dns do not store information about the workspaces (services) anymore. This will reduce the load on that component.

**Changes:**
- Decouple ws-proxy from ws-manager
- Remove GRPC from ws-proxy
- Workspace details are extracted from pods
- Workspace exposed ports are published on the pod itself (annotation `gitpod/exposedPorts`)
- There are no more workspace services

## Related Issue(s)

fixes #6441
fixes #6325

## How to test

Open a workspace. The improvement of speed to expose ports should be noticeable.
Executing `kubectl get svc` should not return any services related to workspaces.

Things to test:
- start a server in the workspace, expose the port and access it
- stop the server and see the port go away
- test local VS code and auto-tunneling
- try `experimentalNetwork: true` in the `.gitpod.yml`

## Release Notes

```release-note
[ws-proxy] Decouple ws-proxy from ws-manager
```
